[WIP] Adding Fixed Width Discretization

hlin117 opened the pull request:
# Update

Please look at #7668 for the current PR.
# Summary

This PR addresses #5778. It adds the Discretizer class to scikit-learn. This PR adds the following:
- [ ] Uniform discretization strategy for 1D arrays
- [x] Uniform discretization strategy for 2D arrays
  - [x] Ability to specify which columns are categorical in a matrix
  - [x] Prevention from passing in invalid categorical column indices
  - [x] Tests for basic discretization capabilities
- [ ] Ability to pass in sparse matrices. (WIP, needs to support previous numpy as scipy versions)
  - [x] Modify `sparsefuncs_fast.pyx` class
  - [x] Use `sparsefuncs.py` for discretization
  - [x] Tests for sparse matrix discretization.
  - [ ] Code review

Documentation
- [x] Parameter and Attribute documentation, in the Sphinx comment format
- [x] Docstring tests
## Doctest Example:

```
>>> # X has two examples, with X[:, 2] as categorical
>>> X = [[0, 1, 0, 4], \
         [6, 7, 1, 5]]
>>> from sklearn.preprocessing import Discretizer
>>> discretizer = Discretizer(n_bins=3, categorical_features=[2])
>>> discretizer.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
Discretizer(...)
>>> discretizer.cut_points_ # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
array([[ 2. , 3. , 4.3333...],
       [ 4. , 5. , 4.6666...]])
>>> # Transforming X will move the categorical features to the last indices
>>> discretizer.transform(X)
array([[0, 0, 0, 0],
       [2, 2, 2, 1]])
```

---
### Original post:

The discretizer should be used as follows. (Just copying from the docstring test.)

```
>>> from sklearn.preprocessing import FixedWidthDiscretizer
>>> X = [0, 1, 2, 3, 4, 5, 6]
>>> discretizer = FixedWidthDiscretizer(n_bins=3)
>>> discretizer.fit(X)
FixedWidthDiscretizer(n_bins=3)
>>> discretizer.cut_points_
array([2., 4.])
>>> discretizer.transform(X)
array([0, 0, 0, 1, 1, 2, 2])
```

A couple of comments and questions:
- This PR is currently failing the build because the tests assume that you can pass in a 2D array into this discretizer. However, I'm limiting the discretizer to only accept 1D arrays. I'm modeling this class off of `LabelEncoder`, which also requires 1D arrays. Does it sound fine that we limit the inputs to only 1D arrays?
- I'm not using a one hot encoding, as @mblondel requested in his original issue in #5778. How do you feel about this, @mblondel?
- `FixedWidthDiscretizer` is a mouthful. Do we have opinions about what else we could name the class? (I'm thinking `KBinsDiscretizer`.)
- Unit tests need to be added.

hlin117 committed:
#5778: Created KBins discretizer

hlin117 committed:
#5778: Refactored code, added doctest

hlin117 committed:
#5778: Fixed doctest, added spacing_ attribute

hlin117 committed:
#5778: Renamed X to y, removed y=None

agramfort commented:
X must be 2D and discretization should be done for each column / feature

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,106 @@
+]
+
+
+class FixedWidthDiscretizer(BaseEstimator, TransformerMixin):
```
I would just use `Discretizer` and later have a `strategy` option in the constructor.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,106 @@
+        self.cut_points_ = None
+        self.spacing_ = None
+
+    def fit(self, y):
```
You need to support 2d `X` here. Also write tests to make sure that this works with both dense and sparse data.

hlin117 replied:
Thanks for the review. Should discretization be done like this?

```
>>> from sklearn.preprocessing import Discretizer
>>> dis = Discretizer(n_bins=3, strategy="uniform", continuous_features=[0, 1, 2])
>>> X = [[0, 1, 4, 0],
         [6, 7, 5, 1]]  # Two examples here, with four features. Last feature is categorical
>>> dis.fit(X)
Discretizer(n_bins=3)
>>> dis.cut_points_
array([[2., 3., 4.3333],
       [4., 5., 4.6666]])
>>> dis.spacing_
array([2., 2., 0.3333])
>>> dis.transform(X)
array([[0, 0, 0],
       [2, 2, 2]])
```

hlin117 replied:
(Is there an update to this, @mblondel?)

mblondel replied:
Sounds good. I would rather use `categorical_features` instead of `continuous_features` to mirror `OneHotEncoder`.

mblondel commented:
> I'm not using a one hot encoding, as @mblondel requested in his original issue in #5778

As suggested by@jnothman, this can be handled easily by a OneHotEncoder in a pipeline. So maybe for separation of concern reasons, it's better to not support one-hot encoding in Discretizer, unless we find out that it is much more efficient to do both in one step.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,106 @@
+
+        self.min_ = y.min()
+        self.max_ = y.max()
+        self.spacing_ = (self.max_ - self.min_) / self.n_bins
```
It seems like it's not necessary to save `spacing` as an attribute: it's not used in `transform` and can be inferred from `cut_points`.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,106 @@
+        self.max_ = y.max()
+        self.spacing_ = (self.max_ - self.min_) / self.n_bins
+        cut_points = (self.min_ + self.spacing_ * i
+                      for i in xrange(1, self.n_bins))
```
I think this can be implemented using `np.linspace`.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,106 @@
+        self.spacing_ = (self.max_ - self.min_) / self.n_bins
+        cut_points = (self.min_ + self.spacing_ * i
+                      for i in xrange(1, self.n_bins))
+        self.cut_points_ = np.array(list(cut_points))
```
Why use generator comprehension if you convert it to a list right after?

hlin117 replied:
The original line was `np.array(list(self.min_ + self.spacing_ * i for i in xrange(1, self.n_bins)))`, which looked like a mouthful. `np.linspace` is a much better implementation though. 

(The implementation used to be using `arange`, but this caused roundoff errors.)

hlin117 committed:
#5778: Renamed FixedWidthDiscretizer to Discretizer

hlin117 committed:
#5778: Allowing Discretizer to accept 2D arrays

hlin117 commented:
Alright @mblondel, the code is ready to be reviewed. 

Couple of comments:

* `strategy` is a lonely parameter right now, because there's only one discretization method. What other discretization methods would you like to see? (One idea is having cut points as inputs, as discussed in #5778.)
* I did my best to avoid for-loops, but alas I have two, <a href="https://github.com/scikit-learn/scikit-learn/pull/5825/files#diff-122fad4f90405b4ac9d6bec08dfc0e61R138">138</a> and line <a href="https://github.com/scikit-learn/scikit-learn/pull/5825/files#diff-122fad4f90405b4ac9d6bec08dfc0e61R171">171</a>. `np.linspace` and `np.searchsorted` both can only accept vectors as inputs, and not matrices. Do you have a suggestion of writing more idiomatic numpy code?

mblondel commented:
I suggest that in this PR you focus on the uniform strategy only. I prefer to have only one strategy but good support for sparse `X`. So for now you can remove the strategy option, we'll add it back later. Please make sure to add tests for sparse data.

In another PR we can add a quantile based approach. Basically you sort the feature values across all samples and choose the cutting points within the sorted values. This will be harder to implement for sparse data and thus needs to be addressed in a separate PR.

Thanks for the good work, you're getting there.

hlin117 committed:
#5778: Removed the 'strategy' parameter

hlin117 committed:
#5778: Added support for sparse matrices csc and csr

hlin117 commented:
Hi @mblondel, thank you for supporting this PR =]

My latest two commits removed the `strategy` parameter, and I added support for `csc` and `csr` sparse matrices. I also added a test case for sparse data. 

Suggestions on how to improve the code is totally welcome. Thanks!

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,178 @@
+    def _check_sparse(self, X, ravel=True):
+        if ravel:
+            return X.toarray().ravel() if sp.issparse(X) else X
+        return X.toarray() if sp.issparse(X) else X
```
Don't densify the sparse matrix. Otherwise there's no point using sparse structures in the first place.

hlin117 replied:
Thanks for checking this, @mblondel. There's only a few places where I call this function, and that's when the passed in array `X` is actually a vector, and I need it to be dense. (Example, setting `self.min_`)

The only situation when I densify a matrix is on line <a href="https://github.com/scikit-learn/scikit-learn/pull/5825/files#diff-122fad4f90405b4ac9d6bec08dfc0e61R176">176</a>, and I don't know how to avoid this call. `discretized` on line 175 is a dense matrix, and you can't use `np.hstack` on a dense/sparse matrix.

This function might get borked at the end of the day. scipy 0.9.0 does not have `csr_matrix.min()`, so that's why the current build is failing.

mblondel replied:
How about CSC format? Does it have a min() method? It's ok to support CSC only for now.

If even CSC doesn't have a min() method in 0.9, you can add a small utility function here:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/sparsefuncs_fast.pyx

mblondel replied:
BTW, it's ok to have two distinct implementations (one for dense and one for CSC) if it makes the implementation clearer.

hlin117 replied:
Neither <a href="https://docs.scipy.org/doc/scipy-0.9.0/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix">csr_matrix</a> nor <a href="https://docs.scipy.org/doc/scipy-0.9.0/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix">csc_matrix</a> have the min or max function implemented in scipy 0.9.0.

I might take your suggestion in adding a utility in `sparsefuncs_fast.pyx`. Thank you!

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,178 @@
+        self
+        """
+
+        X = check_array(X, accept_sparse=self.sparse_formats)
```
CSR is fine if categorical_features="all" but you'll probably need CSC if not because you need feature-wise access to the data. Have a look at the `data`, `indices`, `indptr` attributes of the CSC format. You'll need to access those to filter out continuous features. If categorical_features="all", you will only need `data` so both CSR and CSC should be fine.

BTW, ideally the transformation should preserve sparsity. So zeros should remain zeros.

hlin117 replied:
Sounds good. Thank you for this tip. I'll add in another unit test to test for this.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,178 @@
+
+    def __init__(self, n_bins=2, categorical_features="all"):
+        if n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
```
We usually do input validation in fit.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,178 @@
+    """
+    sparse_formats = ['csr', 'csc']
+
+    def __init__(self, n_bins=2, categorical_features="all"):
```
If categorical_features="all", there is nothing to do right? I think the default should be None.

hlin117 committed:
#5788: Input validation in fit, not constructor. And fixed param options

hlin117 committed:
#5778: Wrote sparse functions and tests for csr_matrix

hlin117 committed:
#5778: Adding sparse modification to Discretizer class

hlin117 committed:
#5778: Renamed the data types

hlin117 committed:
\#5778: Fixing up min and max for csr

\#5788: Resolving csc bugs

hlin117 committed:
\#5778: Fixed issue with integer datasets

\#5788: More fixes to the min and max functions

hlin117 commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -14,12 +15,109 @@ cimport cython
+        col_indices = X.indices
+        np.ndarray[DOUBLE, ndim=1, mode="c"] minimum = np.zeros(n_features)
+        np.ndarray[DOUBLE, ndim=1, mode="c"] data = X.data
+        #DOUBLE[::1] data = X.data
```
When I used `DOUBLE[::1] data = X.data` instead of `np.ndarray[DOUBLE, ndim=1, mode="c"]`, I ended up getting segmentation faults when I ran my code while running `nosetests test_sparsefuncs.py`. Can someone help explain why?

hlin117 commented:
Hi @mblondel, I wrote the code to find the `axis=0` min/max of csr/csc matrices. Can you help review my code?

hlin117 commented on sklearn/utils/sparsefuncs.py:
```diff
@@ -25,6 +26,53 @@ def _raise_error_wrong_axis(axis):
+
+    # TODO: I don't know how to avoid this call to `astype()`;
+    # my test cases don't pass with integers otherwise.
+    X = X.astype(float)
```
Is there any way to avoid calling this? I actually run into the following error if I don't convert my data into floats:

    ValueError: Buffer dtype mismatch, expected 'DOUBLE' but got 'long long'

I usually trigger this error on line <a href="https://github.com/scikit-learn/scikit-learn/pull/5825/files#diff-b538b51a4a930c310a7fea81b377e9bbR57">57</a> of `sparsefuncs_fast.pyx`.

mblondel replied:
This is type of conversion is necessary because the Cython code doesn't do the conversion for us. I would use `np.float64` instead of `float` to be more explicit. This is the type that corresponds to `double`.

hlin117 commented:
I'm currently running into the issue of indexing into sparse matrices. Unfortunately, it seems that scipy sparse matrices are bad with slicing:

<img width="530" alt="screenshot 2015-11-22 13 15 25" src="https://cloud.githubusercontent.com/assets/1865885/11328411/e9fd9ddc-913e-11e5-9184-a995774af336.png">

Though this is the 0.11 version of scipy, I also replicated this bug in my non-local version, 0.16. I'm assuming that it exists for 0.9 as well. I'm currently seeing whether this bug exists in the master branch.

Not sure if any of the other scikit-learn developers have encountered this bug before.

mblondel commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,190 @@
+        # Set the mask for the continuous features in the array
+        self._set_continuous_mask()
+
+        continuous = X[:, self.continuous_mask_]
```
Slicing / masking w.r.t. features is only possible using CSC format. Please accept only CSC format in `check_array` above.

hlin117 replied:
Hi @mblondel. I decided to replace the mask with a list of indices of the continuous features in 021818e. What do you think? Is this a better solution?

hlin117 committed:
#5778: Replaced continuous masks with list indexers

hlin117 committed:
#5778: Changed min and max of csr/csc to use np.float64 conversion

mblondel commented:
The use of searchsorted in transform is problematic, since it seems difficult to avoid a loop.
Maybe we will have to use Cython for transform as well.

An algorithm possibly worth investigating:
1. first scale features to [0, 1] range 
2. round using np.round
3. map to integers

For step 1 I thought we could use MinMaxScaler but it doesn't seem to work on sparse data.
Also, not sure if it's going to be possible to avoid a loop for step 3.

In any case, we need to design a discretization scheme that preserves sparsity (zeros are mapped to zeros). This would mean that features smaller than 0 are mapped to negative numbers.

@jnothman and @amueller might want to pitch in :)

mblondel commented:
> In any case, we need to design a discretization scheme that preserves sparsity

This should be automatically the case for non-negative sparse data. For sparse data including negative values, this might need special care.

hlin117 commented:
@mblondel: How about this idea?

First, let's assume that when a user creates the discretizer, the user will not add more intervals to the class. (The interface doesn't allow this anyways.) Also, let's suppose that we have intervals

`[-infty, x1], [x2, x3], ... [xK <= 0 < x(K+1)]..., [xL, infty]`

I want to represent this in a binary search tree, kind of format. This is how one would discretize an element:

### Upon creation

1. Instantiate a sorted array such that each cell holds both a `key` and `value` with these three rules:
  * The array is sorted by key
  * `xK` is excluded from being a `key`, and each `value` must be at least 1.
  * `np.inf` and `-np.inf` are also keys.

### When discretizing `x`

1. If `x` is within [xK, x(K+1)], return 0.
2. Otherwise, just perform a binary search on this array for some `key=xi` such that `xi <= x < x(i + 1)`. The returned value is the `value` whose key is `xi`. 
  * Observe that the sentinel `np.inf` and `np.inf` values help us map values between `[-infty, x1]` and `[xL, infty]`. 

Notice step 1 takes advantage that sparse matrices have mostly 0's. Also, this isn't too difficult to implement.

I'd like to take a stab at this because

hlin117 commented:
Actually, you could just use `np.searchsorted` on an array that has `xK` taken out. And then just add 1 to the output values. That's the easier way to work this problem without having to code cython.

hlin117 commented:
Sorry, one more afterthought: `np.searchsorted` doesn't work with sparse matrices. The binary search approach might be the best way to solve this problem.

hlin117 committed:
#5778: Replaced cut_points_ attribute with zero_intervals_ and searched_points_

hlin117 committed:
#5778: Added test case, fixed bugs related to new discretization scheme

hlin117 committed:
#5778: Adding cython discretization, need to test

hlin117 committed:
#5778: Preliminary fixes to the binary search

hlin117 committed:
#5778: Resolving dense matrix discretization

hlin117 committed:
#5778: Can now discretize csc_matricies with 2 bins

hlin117 committed:
#5778: Added test case to discretize sparse with 3 bins

hlin117 committed:
#5778: Releasing python gil in cython

hlin117 committed:
#5778: Fixed documentation, added tests for single categorical feature

hlin117 committed:
#5778: Adding setup.py

hlin117 commented:
@mblondel: I rewrote a lot of the code to accommodate the binary search implemented in cython. Can you take a look and tell me what you think?

vene commented:
>  Transforming X will move the categorical features to the last indices

I'm not sure why this should be like this. Do you mean the last index overall, among all the features?

If the output is fed into (1.) an one-hot encoder or (2.) a tree-based model, it would make no difference, right?

If we could do away with this feature, this transformer could just assume that no features are numerical. This will simplify the code and interface. Since this is a transformer and not a classifier, users can easily work around the issue with `FeatureUnion`s. (The continuous and non-continuous data probably come from different sources at some point, so just discretize sooner rather than later.)

What do you think? (ping @jnothman, @mblondel)

(P.S. @mblondel, if this works well with polynomial features, does it mean that fuzzy logic is making a comeback? :stuck_out_tongue:)

vene commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,341 @@
+    max_ : float
+        The maximum value of the input data.
+
+    cut_points_ : array, shape [numBins - 1, n_continuous_features]
```
`n_bins`

vene commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,341 @@
+from __future__ import division
+
+__author__ = "Henry Lin <hlin117@gmail.com>"
```
I think our convention is `# Author: Henry Lin <hlin117@gmail.com>`, not sure how much this matters.

vene commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,341 @@
+
+    Example
+    -------
+    X has two examples, with X[:, 2] as categorical
```
I think you mean "three samples" rather than "two examples"

vene commented:
Could this use the bin edge representation from [`np.histogram`](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.histogram.html) for consistency, or is there some downside?

vene commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,341 @@
+    Attributes
+    ----------
+    min_ : float
+        The minimum value of the input data.
```
for each (non-discrete) column?

hlin117 replied:
Yes, for each non-discrete (non-categorical / continuous) column. I should probably be more explicit about this.

vene commented on sklearn/utils/sparsefuncs.py:
```diff
@@ -25,6 +26,49 @@ def _raise_error_wrong_axis(axis):
         raise ValueError(
             "Unknown axis value: %d. Use 0 for rows, or 1 for columns" % axis)
 
+def csr_csc_min_axis0(X):
```
I'm curious, is there a performance problem with simply calling `X.min(axis=0)` and `X.max(axis=0)`? At first glance the code invoked there seems efficient. Are these functions faster?

hlin117 replied:
Unfortunately, the minimum scipy version that scikit-learn doesn't have the `min` functions for `csr` and `csc` sparse matrices.

vene replied:
Could this be implemented as a backport somehow then?

Also I think you can reduce the amount of code here, because the min of a csc axis 0 is the same as the min of a csr axis 1, and viceversa.

(The scipy implementation uses generic reducers and ufuncs, which might be overkill to backport. But we need to make sure that this is treated as a temporary solution that will be removed when we drop the old scipy.)

hlin117 replied:
I'll agree to working on the backport solution, what would be the necessary steps to get this started?

(I'll also try to reduce the amount of code duplication.)

vene commented on sklearn/preprocessing/tests/test_discretizer.py:
```diff
@@ -0,0 +1,192 @@
+def test_discretizer_1d():
+    pass # TODO
+
+def test_discretizer_bad_n_bins():
```
Please use `from nose.tools import assert_raises`

hlin117 replied:
Addressed in 01b934bee591b41ce32fa6e7fa1c370440983046

vene commented on sklearn/preprocessing/_discretization.pyx:
```diff
@@ -0,0 +1,55 @@
+            else:
+                keyindex = binary_search(value, _search_points)
+                output[i] = keys[keyindex]
+            i += 1
```
isn't `i` invariantly equal to `value_index` here? I can't tell what would make it different.

hlin117 replied:
You're right, I'll change that in the next commit. Thanks.

vene commented on sklearn/preprocessing/_discretization.pyx:
```diff
@@ -0,0 +1,55 @@
+            value = values[value_index]
+            if lower <= value < upper:
+                output[i] = 0
+            else:
```
Instead of a special condition for zeros here, this could be baked into the `keys` array.

`keys` could be designed to go from `[-k, ..., 0, ... +z]` ensuring that zeros map to zeros.
(The only important reasons why zeros map to zeros is so that we can safely iterate only over nonzeros when given a sparse matrix as input.)

vene commented on sklearn/preprocessing/discretization.py:
```diff
@@ -0,0 +1,341 @@
+        return cut_points
+
+    @property
+    def cat_features_(self):
```
:cat2:

hlin117 replied:
Hmm, do you think this function should be renamed as `categorical_features_`? I guess `cat_features` is an abbreviation, which might not be very ideal.

vene commented:
1. With the current implementation, `0` is a special value bin, but all other bins are labeled 1, 2, ...  I think that is a bit confusing, @mblondel suggested outputting negative bins.  Is there a downside to that in your applications, @hlin117?

2. @hlin117, your code is very well written, I'm just concerned that there is *too much* code for such a simple function. So, in a way, most of my comments are about reducing the size of this contribution and simplifying it, to make it more manageable and maintainable. 

In this direction: have you tried out the rounding-based approach suggested by @mblondel above? How does it compare?

hlin117 commented:
Hey @vene,

1. I'm confused about @mblondel's algorithm above. He mentioned the following steps:

> 
> An algorithm possibly worth investigating:
> 1. first scale features to [0, 1] range 
> 2. round using np.round
> 3. map to integers
> 
> For step 1 I thought we could use MinMaxScaler but it doesn't seem to work on sparse data.
> Also, not sure if it's going to be possible to avoid a loop for step 3.
> 
> In any case, we need to design a discretization scheme that preserves sparsity (zeros are mapped to zeros). This would mean that features smaller than 0 are mapped to negative numbers.

Suppose we have the following three continuous features (and imagine that the data were sparse, like a sparse array), and I wanted to split this feature into `k=3` bins:
`[-1, 0, 1]`
Rescaling to `[0, 1]` would yield
`[0, 0.5, 1]`
Rounding this would yield
`[0, 0, 1]`
And splitting this into `k=3` bins doesn't make sense to me.

Am I misinterpreting what Mathieu's algorithm is supposed to be doing? I'd like to consider other, more simple, implementations that don't require a binary search.

2. I totally agree that this is too much code for this PR.

hlin117 commented:
>>Transforming X will move the categorical features to the last indices

>I'm not sure why this should be like this. Do you mean the last index overall, among all the features?

I'm imagining the discretization to happen like this. Suppose we have the following input matrix.

```
X = [[10, 1, 3, 4, 5]
     [0, -1, 0, 1, 0]
     [1, 0, 1, 1, 0]]
```

Let's assume that the second index (the column with `[1, -1, 0]`) is the continuous column and everything else is categorical. Then discretization with `k=3` would yield

```
X = [[2, 10, 3, 4, 5]
     [1, 0, 0, 1, 0]
     [0, 1, 1, 1, 0]]
```

So, the columns with `10, 3, 4, 5` are towards the last indices of the array.

This behavior is the same as <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html">pd.get_dummies</a> and `OneHotEncoder`. 

```
>>> from sklearn.preprocessing import OneHotEncoder
>>> 
>>> # Second column (starting with 666) is categorical
>>> arr = [[1, 666, 3, 4, 5],
>>>        [1, 667, 3, 4, 5],
>>>        [2, 668, 3, 4, 5]]
>>> 
>>> OneHotEncoder(categorical_features=[1]).fit_transform(arr).toarray()
array([[ 1.,  0.,  0.,  1.,  3.,  4.,  5.],
       [ 0.,  1.,  0.,  1.,  3.,  4.,  5.],
       [ 0.,  0.,  1.,  2.,  3.,  4.,  5.]])
```

vene commented:
Regarding rounding, I assume @mblondel's idea went something like:

(for dense data) 
 * scale X into [0, k]
 * round

which seems like it would give you k + 1 bins. I'm not sure if this is right, nor how to make it work for sparse data, but it does make sense to me.

Regardless the other issue, but I don't understand why you would need to change anything in the discrete columns. And if we don't change the discrete columns, we can remove all the code that deals with discrete columns, simplifying this greatly (including the :cat: method :) )

hlin117 commented:
> Regarding rounding, I assume @mblondel's idea went something like:
> 
> (for dense data)
> 
> * scale X into [0, k]
> * round
> which seems like it would give you k + 1 bins. I'm not sure if this is right, nor how to make it work for sparse data, but it does make sense to me.
Ah, that's clever. However, I think there are cases when this wouldn't work.

For example, if we have `[-1, 0, 1]`, and we want `k=3` bins, then `[-1, 0, 1]` becomes `[0, 1.5, 3]` (by scaling into `[0, k]`), and after rounding, we would have `[0, 1, 3]`. So we unintentionally add another class in there.

I think scaling into `[0, k-1]` is the correct solution, but - like you mentioned - it won't work for sparse data. 

If we scale into `[0, k-1]`, we would have `[0, 1, 2]`. The zero's in the sparse matrix won't be mapped to zero's in the transformed matrix.

hlin117 commented:
> Regardless the discrete columns: I don't understand why you would need to change anything in the discrete columns. And if we don't change the discrete columns, we can remove all the code that deals with discrete columns, simplifying this greatly (including the :cat: method :) )

I just shifted the discrete (categorical) columns to the end of the matrix, just because - implementation wise - this simplified my code. I'd imagine that this is the case with OneHotEncoder too (just simplifying implementation).

But on the other hand, the OneHotEncoder and `pd.get_dummies` adds more features to the feature matrix. In my case, it shouldn't be terribly difficult to keep things in the same columns.

hlin117 committed:
Addressing minor changes from @vene

hlin117 committed:
Fixing build by using embedded six package

hlin117 commented:
It looks like travis's python3 build is failing because it can't find the `_discretization` module. I can replicate this error on my computer, but I don't know how to fix it. 

I reckon that my configurations in [setup.py](https://github.com/hlin117/scikit-learn/blob/kbins/sklearn/preprocessing/setup.py) are incorrect, but I don't understand enough Cython to know. If someone can diagnose why the build is failing, that would be great.

hlin117 commented:
Alright, I think I understand the scaling scheme @mblondel was mentioning earlier.

I imagine the scaling to go like this. Given some feature `X` and our chosen number of bins, `k`:
1. Let `X_trans = X * k / (X.max() - X.min())`. Observe that `X_trans` is a dataset of width `k`.
2. `epsilon = abs(X_trans.min() - round(X_trans.min()))`.
3. `X_trans[X_trans != 0] += epsilon`. We do this operation because it works with sparse matrices. 
4. `X_trans = np.round(X_trans)`
5. `X_trans.eliminate_zeros()`, for sparse matrices
6. `X_trans` now contains all of the discretized features.

Like @vene said, we need to encourage code maintainability, and I think this 6 step process would be faster and much less code than the current implementation. It also works for sparse matrices, which is essentially our blocker here.

Please let me know if there is any interest in continuing this PR.

vene commented:
@hlin117 I'm guessing you need the epsilon to make sure that the smallest non-zero values get discretized to non-zero? I'm guessing you could just add 1 in the first step instead, right? (basically rescale all non-zeros to lie in (1, 1+k) instead of (0, k).

Indeed this was my understanding of @mblondel's suggestion. We need to check two things now:

 - [ ] how different this is from the explicit discretization (ideally both in approximation error and in downstream utility when running some classifier on the result)
 - [ ] how it compares in terms of runtime.

Also if the rounding approach does work well and is fast, we will need to document the code really clearly, because it's pretty counterintuitive.

amueller commented:
Any news on this?

hlin117 commented:
I'll be happy to continue this PR, if there's support for it.

jnothman commented:
I think there is interest.

In the case of sparse matrices, you're only going to be able to maintain sparsity if you engineer 0 to remain 0. I'm confused whether that's what you're doing with your `epsilon`.

I think you're also missing a `clip` stage in your processing, as the transformed data may have outliers relative to the fit ranges.

hlin117 commented:
I think if we were to move forward with this, I would be okay with abandoning my code here, and starting afresh.

hlin117 commented:
@jnothman 
> I think you're also missing a clip stage in your processing, as the transformed data may have outliers relative to the fit ranges.

I think this data preprocessing should be up to the user. Some users may choose not to clip off these outliers, and it's not really clear how to define what is an outlier, and what is not.

jnothman commented:
Outliers are things out of range of the training data and therefore its
bins. These should be handled by the transformer.

On 14 October 2016 at 17:53, Henry Lin <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman>
>
> I think you're also missing a clip stage in your processing, as the
> transformed data may have outliers relative to the fit ranges.
>
> I think this data preprocessing should be up to the user. Some users may
> choose not to clip off these outliers, and it's not really clear how to
> define what is an outlier, and what is not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5825#issuecomment-253723540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61dgwRf91K5-x7ufIk_n8S7aAZnNks5qzycHgaJpZM4Giwch>
> .
>

hlin117 commented:
> Outliers are things out of range of the training data and therefore its
bins. These should be handled by the transformer.

Ah true. I wasn't thinking about the testing set, only the training set. Thank you.

hlin117 commented:
Please look at 7668. It contains the much more straightforward algorithm. (And much much much less code.) Thanks!
